src/prefect/client/utilities.py

```diff
@@ -5,32 +5,31 @@
 # This module must not import from `prefect.client` when it is imported to avoid
 # circular imports for decorators such as `inject_client` which are widely used.
 
+from collections.abc import Awaitable, Coroutine
 from functools import wraps
-from typing import (
-    TYPE_CHECKING,
-    Any,
-    Awaitable,
-    Callable,
-    Coroutine,
-    Optional,
-    Tuple,
-    TypeVar,
-    Union,
-    cast,
-)
-
-from typing_extensions import Concatenate, ParamSpec
+from typing import TYPE_CHECKING, Any, Callable, Optional, Union
+
+from typing_extensions import Concatenate, ParamSpec, TypeIs, TypeVar
 
 if TYPE_CHECKING:
     from prefect.client.orchestration import PrefectClient, SyncPrefectClient
 
 P = ParamSpec("P")
-R = TypeVar("R")
+R = TypeVar("R", infer_variance=True)
+
+
+def _current_async_client(
+    client: Union["PrefectClient", "SyncPrefectClient"],
+) -> TypeIs["PrefectClient"]:
+    from prefect._internal.concurrency.event_loop import get_running_loop
+
+    # Only a PrefectClient will have a _loop attribute that is the current loop
+    return getattr(client, "_loop", None) == get_running_loop()
```

Comment:
I may have used the wrong special form here. This type narrowing function returns a `TypeIs` construct but now that I reread this it should have been a `TypeGuard` instead.

That's because when this function returns False in two different cases:

1. the client is a `SynPrefectClient` instance.
2. the client is a `PrefectClient` instance, attached to a different loop.

The `TypeIs` construct acts like a `isinstance()` check, where the type checker can infer that if False is returned then the client must be a `SyncPrefectClient`. But that's not correct for case 2.

A `TypeGuard` function explicitly doesn't make promises about types when False is returned and so fits this function much better.

It is a technically moot point **in how this function is used in this module** but an important distinction if you were to use this as an example 🤦.

TLDR: 

- use `TypeIs[R]` when True means _this is of type `R`_ and False means _this is not of type `R`_
- use `TypeGuard[R]` when True means _this meets conditions that tell me this is of type `R`_ but False means _uh, I am not sure what this is and I don't care 🤷_.